src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -672,13 +672,23 @@ impl RustcTargetData {
         let host_info = TargetInfo::new(config, requested_kinds, &rustc, CompileKind::Host)?;
         let mut target_config = HashMap::new();
         let mut target_info = HashMap::new();
-        for kind in requested_kinds {
-            if let CompileKind::Target(target) = *kind {
+        // Note that we cannot just rely on getting kinds for all units as the units aren't built yet (TODO CHECK THIS IS TRUE?)
```

Comment:
Hm this may pose a bit of a chicken-and-egg problem where we may not be able to precalculate all target information ahead of time like we do today. This'll definitely  need to iterate over the entire dependency graph to discover targets since any package could introduce a new target I think.

Reply:
Do you think lazily filling the two `HashMaps` in the functions that actually access them could make sense?

Reply:
Hmm… Actually, I thought I'd just go ahead and implement this, but it turns out it's a much bigger refactoring than I initially thought. I've sent a WIP PR at https://github.com/rust-lang/cargo/pull/9049 ; if you think that's a good way to go I can finish it (already went up to the borrowck phase), but I'm starting to feel like this is not good design.

Another idea I had may be: add a function to `RustcTargetData` to manually request loading of more target information, and then call it from the places that may introduce new targets. What do you think about this other idea?

Reply:
I don't have much of a strong preference for how this is fixed myself, I mostly just wanted to point out that this was a problem that needed fixing with this implementation.